File: PA6/main/stack.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdbool.h>
#include "stack.h"

// Edit only after this line //

/**
 * current size of stack
 *
 * @param s Stack pointer
 * @return size of Stack
 */
int currentSize(Stack* s) {
  return s->top + 1;
}

/**
 * Helper function to shrink capacity of stack according to shrinkFactor
 *
 * @param s Stack pointer
 */
void shrinkCapacity(Stack* s) {
  Element** newArray = (Element**) malloc(s->capacity * sizeof(Element*));
  for (int i = 0; i < currentSize(s); i++) {
    newArray[i] = s->elements[i];
  }
  free(s->elements);
  s->elements = newArray;
}

/**
 * pop function for stack
 *
 * @param s Stack pointer
 * @return popped top element
 */
Element* pop(Stack* s) {
  // stack empty
  if (isEmpty(s)) {
    return NULL;
  }
  Element* popped = s->elements[s->top];
  s->top--;
  // shrinkFactor triggered
  if (s->shrinkFactor != 0 && currentSize(s) < (int) (getCapacity(s)
      * s->shrinkFactor)) {
    s->capacity = s->capacity * .5;
    shrinkCapacity(s);
  }
  return popped;
}

/**
 * pop multiple elements from stack
 *
 * @param s Stack pointer
 * @param k number of elements to pop
 * @return popped elements as an array
 */
Element** multiPop(Stack* s, int k) {
  if (isEmpty(s)) {
    return NULL;
  }
  // pop all
  Element** arr;
  if (k > currentSize(s)) {
    arr = (Element**) malloc(currentSize(s) *
      sizeof(Element*));
    for (int i = 0; i < currentSize(s) - 1; i++) {
      arr[i] = pop(s);
    }
    return s->elements;
  } else {
    arr = (Element**) malloc(k * sizeof(Element*));
    for (int i = 0; i < k; i++) {
      arr[i] = pop(s);
    }
    return arr;
  }
}

/**
 * Utility function: expandCapacity for push
 *
 * @param s Stack pointer
 */
void expandCapacity(Stack* s) {
  Element** newArray = (Element**) malloc(s->capacity * sizeof(Element*));
  for (int i = 0; i < currentSize(s); i++) {
    newArray[i] = s->elements[i];
  }
  free(s->elements);
  s->elements = newArray;
}

/**
 * push function for stack
 *
 * @param s Stack pointer
 * @param k int key
 * @param v string value
 * @return push successful or not
 */
bool push(Stack* s, int k, char* v) {
  // stack full
  if (isFull(s)) {
    return false;
  }
  Element* item = malloc(sizeof(struct Element));
  item->key = k;
  item->value = v;
  s->elements[++s->top] = item;
  // growFactor triggered
  if (s->growthFactor != 0 && currentSize(s) > (int) (getCapacity(s)
      * s->growthFactor)) {
    s->capacity += s->capacity;
    expandCapacity(s);
  }
  return true;
}

/**
 * push unique function for stack, push only if top element not equal to
 * params: key and value
 *
 * @param s Stack pointer
 * @param k int key
 * @param v string value
 * @return push successful or not
 */
bool pushUnique(Stack* s, int k, char* v) {
  // stack full
  if (isFull(s)) {
    return false;
  }
  if (s->elements[s->top]->key != k || s->elements[s->top]->value != v) {
    push(s,k,v);
    return true;
  }
  return false;
}

/**
 * reverse the elements in stack
 *
 * @param s Stack pointer
 */
void reverse(Stack* s){
  // reverse stack (array)
  for (int i = 0; i < currentSize(s) / 2; i++) {
    Element* tmp = s->elements[i];
    s->elements[i] = s->elements[currentSize(s) - 1 - i];
    s->elements[currentSize(s) - 1 - i] = tmp;
  }
}

/**
 * peek function for stack
 *
 * @param s Stack pointer
 * @return top element of Stack
 */
Element* peek(Stack* s) {
  // stack empty
  if (isEmpty(s)) {
    return NULL;
  }
  return s->elements[s->top];
}

/**
 * search for element from top of stack
 *
 * @param s Stack pointer
 * @param k int key
 * @param v string value
 * @return distance to searched element from top of Stack
 */
int search(Stack* s, int k, char* v) {
  // stack empty
  if (isEmpty(s)) {
    return -1;
  }
  int distance = 1;
  int i = s->top;
  while (i != -1) {
    // found
    if (s->elements[i]->key == k && s->elements[i]->value == v) {
      return distance;
    }
    distance++;
    i--;
  }
  return -1;
}

/**
 * get current capacity of stack
 *
 * @param s Stack pointer
 * @return capacity of Stack
 */
int getCapacity(Stack* s) {
  return s->capacity;
}

/**
 * check if stack is full
 *
 * @param s Stack pointer
 * @return Stack full or not
 */
bool isFull(Stack* s) {
  // stack is always growing
  if (s->growthFactor != 0) {
    return false;
  }
  return s->top == s-> capacity - 1;
}

/**
 * clear the stack
 *
 * @param s Stack pointer
 */
void clear(Stack *s) {
  s->top = -1;
}

/**
 * clean the stack
 *
 * @param s Stack pointer
 */
void cleanStack(Stack *s) {
  for(int i = 0; i < s->capacity; i++) {
    free(s->elements[i]);
  }
    free(s->elements);
    free(s);
}

/**
 * check if stack empty
 *
 * @param s Stack pointer
 * @return whether Stack empty
 */
bool isEmpty(Stack* s) {
  return s->top == -1;
}

/**
 * Stack with fixed capacity, cotr
 *
 * @param cap Stack capacity
 */
Stack* makeStack(int cap) {
  Stack* stack = (struct Stack*) malloc(sizeof(struct Stack));
  stack->top = -1;
  stack->capacity = cap;
  stack->growthFactor = 0;
  stack->shrinkFactor = 0;
  stack->dynamic = false;
  stack->elements = (Element**) malloc(stack->capacity * sizeof(Element*));
  return stack;
}

/**
 * Stack with fixed capacity that can grow, cotr
 *
 * @param cap Stack capacity
 * @param growF Stack growth factor
 */
Stack* makeStackG(int cap, float growF) {
  Stack* stack = (struct Stack*) malloc(sizeof(struct Stack));
  stack->top = -1;
  stack->capacity = cap;
  stack->growthFactor = growF;
  stack->shrinkFactor = 0;
  stack->dynamic = true;
  stack->elements = (Element**) malloc(stack->capacity * sizeof(Element*));
  return stack;
}

/**
 * Stack with fixed capacity that can grow and shrink, cotr
 *
 * @param cap Stack capacity
 * @param growF Stack growth factor
 * @param shrinkF Stack shrink factor
 */
Stack* makeStackGnS(int cap, float growF, float shrinkF) {
  Stack* stack = (struct Stack*) malloc(sizeof(struct Stack));
  stack->top = -1;
  stack->capacity = cap;
  stack->growthFactor = growF;
  stack->shrinkFactor = shrinkF;
  stack->dynamic = true;
  stack->elements = (Element**) malloc(stack->capacity * sizeof(Element*));
  return stack;
}

File: PA4/src/main/java/Stack.java
import java.util.EmptyStackException;
import java.util.*;

class Stack<E>
{
    float loadFactor;
    float shrinkFactor;
    int top; // Index of the top element
    E arr[];
    int capacity;

    Stack(int capacity)
    {
        this.capacity = capacity;
        this.loadFactor = 0;
        this.shrinkFactor = 0;
        this.arr = (E[]) new Object[this.capacity];
        this.top = -1;
    }

    Stack(int capacity, float loadFactor)
    {
        this.capacity = capacity;
        this.loadFactor = loadFactor;
        this.shrinkFactor = 0;
        this.arr = (E[]) new Object[this.capacity];
        this.top = -1;
    }

    Stack(int capacity, float loadFactor, float shrinkFactor)
    {
        this.capacity = capacity;
        this.loadFactor = loadFactor;
        this.shrinkFactor = shrinkFactor;
        this.arr = (E[]) new Object[this.capacity];
        this.top = -1;
    }

    boolean isEmpty()
    {
        // return this.top == 0;
        return this.currentSize() == 0;
    }

    int getCapacity(){
        return this.capacity;
    }

    void push(E x) throws StackOverflowError, NullPointerException
    {
        // Stack can't grow (full)
        if (this.isFull()) {
            throw StackOverflowError();
        }
        // push null
        if (this.x == null) {
            throw NullPointerException();
        }
        // add element to top of Stack
        this.arr[this.top + 1] = x;
        this.top++;
        // handles load factor
        if (this.loadFactor != 0 && this.currentSize() > this.capacity * this.loadFactor) {
            this.capacity *= 1 + this.loadFactor;
        }
    }

    E pop() throws EmptyStackException
    {
        // stack empty
        if (this.isEmpty()) {
            throw EmptyStackException();
        }
        E tmp = this.arr[this.top];
        this.arr[this.top] = null;
        this.top--;
        // handles shrink factor
        if (this.shrinkFactor != 0 && this.currentSize() < this.capacity * this.shrinkFactor) {
          this.capacity *= this.shrinkFactor;
        }
        return tmp;
    }

    E peek() throws EmptyStackException{
        // stack empty
        if (this.isEmpty()) {
            throw new EmptyStackException();
        }
        return this.arr[this.top]
    }

    void clear(){
        // empty Stack
        if (this.isEmpty()) {
            return;
        }
        // loop through array, set to null (remove), decrement top
        for (int i = 0; i < this.arr.length; i++) {
            this.arr[i] = null;
            this.top--;
        }
    }

    boolean isFull(){
        return this.currentSize() == this.capacity;
    }

    int currentSize(){
        int counter = 0;
        while (this.arr[counter] != null) {
            counter++;
        }
        return counter;
    }

    E[] multiPop(int k) throws EmptyStackException{
        // stack empty
        if (this.isEmpty()) {
            throw EmptyStackException();
        }
        // k > size
        if (k > this.currentSize()) {
            return this.arr[];
        }
        E[] popArray = (E[]) new Object[k];
        // pop k elements
        for (int i = 0; i < k; i++) {
            popArray[i] = this.arr[this.top].pop();
            this.top--;
        }
        return popArray;
    }

    void multiPush(E[] arr) throws StackOverflowError, NullPointerException{
        // Stack can't grow (full)
        if (this.isFull()) {
            throw StackOverflowError();
        }
        // loop through param arr
        for (int i = 0; i < arr.length; i++) {
            this.push(arr[i]);
        }
    }

    void reverse() throws EmptyStackException{
        // stack empty
        if (this.isEmpty()) {
            throw new EmptyStackException();
        }
        // reverse stack (array)
        for (int i = 0; i < arr.length / 2; i++) {
            E tmp = arr[i];
            arr[i] = arr[arr.length - 1 - i];
            arr[arr.length - 1 - i] = tmp;
        }
    }

    boolean pushUnique(E x) throws StackOverflowError, NullPointerException{
        // Stack can't grow (full)
        if (this.isFull()) {
            throw StackOverflowError();
        }
        // push null
        if (this.x == null) {
            throw NullPointerException();
        }
        // param x is unique
        if (x != this.arr[this.top]) {
            this.push(x);
            return true;
        }
        // not a success
        return false;
    }

    int search(E x) throws EmptyStackException{
        // stack empty
        if (this.isEmpty()) {
            throw new EmptyStackException();
        }
        int distance = 1;
        int i = top;
        while (i != 0) {
            // found
            if (arr[i] == x) {
                return distance;
            }
            distance++;
            i--;
        }
        return -1;
    }

}
